src/Microsoft.AspNet.OData.Shared/Builder/DerivedTypeConstraintSet.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.OData.Edm.Csdl;
+
+namespace Microsoft.AspNet.OData.Builder
+{
+    /// <summary>
+    /// Set of derived type constraints. 
+    /// </summary>
+    public class DerivedTypeConstraintSet : HashSet<Type>
+    {
+        /// <summary>
+        /// Initializes a DerivedTypeConstraintSet instance without setting a base type.
+        /// </summary>
+        public DerivedTypeConstraintSet()
+        {
+            Location = EdmVocabularyAnnotationSerializationLocation.Inline;
+        }
+
+        /// <summary>
+        /// Initializes a DerivedTypeContraintSet instance with a base type specified.
+        /// </summary>
+        /// <param name="baseType">CLR type of the base class.</param>
+        public DerivedTypeConstraintSet(Type baseType)
+            : base()
+        {
+            ClrBaseType = baseType;
+        }
+
+        /// <summary>
+        /// CLR Type of the base class. 
+        /// </summary>
+        public Type ClrBaseType { get; set; }
+
+        /// <summary>
+        /// Location of the derived type constraint annotation. By default, it will be inline.
+        /// </summary>
+        public EdmVocabularyAnnotationSerializationLocation Location { get; set; }
+
+        /// <summary>
+        /// Add TDerivedType to the set of derived type constraints.
+        /// </summary>
+        /// <typeparam name="TDerivedType">Derived Type.</typeparam>
+        /// <returns>Updated DerivedTypeConstraint set.</returns>
+        public DerivedTypeConstraintSet AddConstraint<TDerivedType>()
+        {
+            ValidateAndAddSingleConstraint(typeof(TDerivedType));
+            return this;
+        }
+
+        /// <summary>
+        /// Add the derived type constraints to the set. 
+        /// </summary>
+        /// <param name="derivedTypes">Derived types to be added to the set.</param>
+        public void AddConstraints(IEnumerable<Type> derivedTypes)
+        {
+            if (!derivedTypes.Any())
+            {
+                ValidateAndAddSingleConstraint(ClrBaseType);
+                return;
+            }
```

Comment:
if the input is empty, add the ClrBaseType ? Why don't we have this strange logic?

Reply:
I had two options; One was to throw an error and the other was to consider that as an indication to allow only base types. Consider that this is called from HasDerivedTypeConstraints(). What should be the behavior?

---
In reply to: [394506704](https://github.com/OData/WebApi/pull/1977#discussion_r394506704) [](ancestors = 394506704)